internal/vcs/git/commits.go

```
@@ -387,7 +390,7 @@ func CommitCount(ctx context.Context, repo api.RepoName, opt CommitsOptions) (ui
 }
 
 // FirstEverCommit returns the first commit ever made to the repository.
-func FirstEverCommit(ctx context.Context, repo api.RepoName) (*gitdomain.Commit, error) {
+func FirstEverCommit(ctx context.Context, repo api.RepoName, checker authz.SubRepoPermissionChecker) (*gitdomain.Commit, error) {
```

I think we can special case this so that it can only be called by an internal actor. If it's called by an actual user we should return an error. 

@coury-clark I assume this is only used internally by code insights?

---

That implementation sounds fine to me, and yeah - this is afiak only used by the background Insights workers at the moment.

---

What's the best way to tell if the caller is an internal actor vs. an actual user? And what type of error would be best to return? Something that indicates no access? A `RevisionNotFound` error?

---

I *think* the answer is something like this:
``` go
actor.FromContext(ctx).IsInternal()
```

https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+IsInternal%28%29&patternType=literal

As far as errors, `RevisionNotFound` works, or some kind of `UnauthorizedRevision` error since this one is pretty unambigious which revision it refers to.

---

What about this case? https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/insights/query/query_executor.go?L56:23
Looks like it's called in a query executor? Not sure if an internal actor is being used there or not.

---

@camdencheek tagging you for my above question. Will this actor always be internal or no?

---

@coury-clark tagging you for my above question. Will this actor always be internal or no?

---

This particular instance will be used with the user context (I forgot we just added this recently). We can of course change this to use internal actor, but I wonder if that will somehow break the constraints you are building for commit filtering.

We have said that Code Insights isn't equipped to work with sub-repo permissions in the past, and we can still take that stance here. Having said that, this looks like a case we didn't fully consider so our team may need to go digest what this means. That isn't to say this is a blocker for you, just that you can leave it as-is (with user context) and as long as `FirstEverCommit` returns an error for a non-visible commit to the user, we won't be leaking anything.

---

So are we cool with leaving this as-is, then? And maybe just disabling the sub-repo permissions checker in the case that the actor is internal?

---

Yes, that should be fine